src/leiningen/ring/uberwar.clj

```diff
@@ -13,25 +17,49 @@
     (get-cp project)
     (->> (:library-path project) io/file .listFiles (map str))))
 
-(defn contains-entry? [^java.io.File file ^String entry]
-  (with-open [jar-file (JarFile. file)]
-    (some (partial = entry)
-          (map #(.getName ^JarEntry %)
-               (enumeration-seq (.entries jar-file))))))
-
 (defn jar-dependencies [project]
   (for [pathname (get-classpath project)
-        :let [file (io/file pathname)
-              fname (.getName file)]
-        :when (and (.endsWith fname ".jar")
-                   ;; Servlet container will have it's own servlet-api impl
-                   (not (contains-entry? file "javax/servlet/Servlet.class")))]
-    file))
-
-(defn jar-entries [war project]
+        :when (str/ends-with? pathname ".jar")]
+    (io/file pathname)))
+
+(def ^:private javax-servlet? #{"javax/servlet/Servlet.class"})
+
+(defn- pom-properties? [file-name]
+  (str/ends-with? file-name "pom.properties"))
+
+(defn- read-jar-pom-properties [^String full-path]
+  (with-open [in (.openStream (URL. full-path))]
+    (->> (doto (Properties.) (.load in))
+         (into {}))))
+
+(defn- jar-name-from-pom-properties
+  [jar-path pom-properties-jar-path]
+  (let [pom-properties-full-path (str "jar:file:"
+                                      jar-path "!/"
+                                      pom-properties-jar-path)
+        {:strs [groupId artifactId version]} (read-jar-pom-properties
+                                               pom-properties-full-path)]
+    (when (and groupId artifactId version)
+      (str groupId \- artifactId \- version ".jar"))))
+
+(defn- war-path-for-jar [^File jar]
+  (with-open [jar-file (JarFile. jar)]
+    (let [javax&pom (->> (.entries jar-file)
+                         enumeration-seq
+                         (map (fn [^JarEntry je] (.getName je)))
+                         (filter (some-fn javax-servlet? pom-properties?)))] ;; 2 elements maximum
+      ;; Servlet container will have it's own servlet-api impl
+      (when-not (some javax-servlet? javax&pom)
+        (str "WEB-INF/lib/"
+             (or (some->> (first javax&pom)
+                          (jar-name-from-pom-properties (.getAbsolutePath jar)))
+                 ;; fallback to using the original jar name (best-effort approach)
+                 (.getName jar)))))))
+
+(defn- populate-war-with-dependent-jars
+  [war project]
   (doseq [jar-file (jar-dependencies project)]
-    (let [dir-path (.getParent jar-file)
-          war-path (war/in-war-path "WEB-INF/lib/" dir-path jar-file)]
+    (when-let [war-path (war-path-for-jar jar-file)]
```

Comment:
Excluding dependencies that contain the javax.servlet class was already the case (see `jar-dependencies`), before I ever laid hands on this project. The `;; Servlet container will have it's own servlet-api impl` comment is not mine - if you look for it in the existing codebase, you'll see that right under it there is a `(not (contains-entry? ...` expression.  So the answer to your question is no - I'm doing that because I don't want to affect what jars are included, only their names.